File: SAtools-test/OneKP-NCBI-Merge/scripts/2_2.Rename-NCBI.pl
#!/usr/bin/perl -w 
# Script: Rename-NCBI.pl
# Usage: perl -w 2_2.Rename-NCBI.pl <infile.fa> <outfile.fa>
# Update date: 2019/08/13
# Author: Zhuofan Zhang

# FORMAT:
# >lcl|NC_XXXXXX.X_[x]+_NP_YYYYYY.Y_Z-Species [..] [..] [[..]...]

use strict;
# use Data::Dumper;
@ARGV == 3 or die "Usage: perl -w 2_2.Rename-NCBI.pl <infile.fa> <outfile.fa> <NCBIid2Species.list>";
my ($ifile,$ofile,$nlist) = @ARGV;
my %id2sp;

open LIST,"<$nlist" or die "$!";
open IFILE,"<$ifile" or die "$!";
open OFILE,">$ofile" or die "$!";

# Load NCBI-ID to SPECIES-NAME list
while(<LIST>)
{
    chomp;
    my @line = split "\t";
    if($line[1]=~ m/^NC_/)
    {
       # Because my stat-file is from Windows, removing the '\r' is needed.
       $line[1] =~ s/\.[0-9]+//g;
       $line[1] =~ s/\r//g;
       $id2sp{$line[1]} = $line[0];
    }

}

# print OFILE Dumper(\%id2sp);

# Rename the sequences
while(<IFILE>)
{
    if(m/^>/)
    {
       chomp;
       $_ =~ m/NC_[0-9]+/;
       # print "$NCID\n";
       $_ =~ m/ \[.*/;
       print OFILE "$`-$id2sp{$NCID}$&\n";
    }
    else
    {    
         print OFILE;
    }
}

File: SAtools-test/OneKP-NCBI-Merge/scripts/2_2.Rename-NCBI.py
# -*- coding: utf-8 -*-
# Script: Rename-NCBI.py
# Description: This is the second version script for processing the NCBI_FASTA.
#              Noted that the input-FASTA has the header format:
#              ">lcl|[NC_ID]_xxx_[NP_ID]-[Species_Name]"
#              and this script will add [taxid] to it and move out those sequences
#              that can't map to proper ID.
# Update date: 2019/09/06
# Author: Zhuofan Zhang

import argparse
import re

parser = argparse.ArgumentParser()
parser.add_argument('-dict',help="species2taxid CSV file.")
parser.add_argument('-i',help="input fasta file.")
parser.add_argument('-o',help="output fasta file.")
args = parser.parse_args()

# As the Dict is not big, read it into memory once.
Species2Taxid = {}
with open(args.dict,'r') as ifile:
    for line in ifile.readlines():
        species,taxid = line.split(',')
        species = species.replace(' ','_')
        Species2Taxid[species]=taxid

with open(args.i,'r') as ifile:
    with open(args.o,'w+') as ofile:
        seq_SPpattern = re.compile(r'-(?P<seq_sp>[a-zA-Z_\.]+) ')
        rline = ifile.readline()
        Filt_No = False
        while(rline):
          if(rline[0]=='>'):
            seq_sp = seq_SPpattern.search(rline).group('seq_sp')
            if(seq_sp in Species2Taxid.keys()):
                Filt_No = True
                wpart = '{}'.format(Species2Taxid[seq_sp].strip())
                wline = re.sub(r'-(?P<seq_sp>[a-zA-Z_\.]+) ','-{} '.format(wpart),rline)
                #print(wpart)
                ofile.write(wline)
            else:
                Filt_No = False
          elif(Filt_No):
            ofile.write(rline)
          rline = ifile.readline()

